src/jennifer/config.cr

```diff
@@ -4,139 +4,157 @@ require "logger"
-      @@retry_delay = 1.0
-
-      @@logger = Logger.new(STDOUT)
-      @@logger.not_nil!.level = Logger::DEBUG
-      @@logger.not_nil!.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
-        io << datetime << ": " << message
+      def logger=(value)
+        @logger = value
       end
-    end
 
-    reset_config
+      def validate_config
+        raise Jennifer::InvalidConfig.new("No adapter configured") if adapter.empty?
+        raise Jennifer::InvalidConfig.new("No database configured") if db.empty?
+      end
 
-    def self.logger
-      @@logger.not_nil!
-    end
+      def from_uri(db_uri : String)
+        begin
+          from_uri(URI.parse(db_uri))
+        rescue e
+          self.logger.error("Error parsing database uri #{db_uri}")
+        end
+      end
 
-    def self.logger=(value)
-      @@logger = value
-    end
+      def from_uri(uri : URI)
+        @adapter = uri.scheme.to_s if uri.scheme
+        @host = uri.host.to_s if uri.host
+        @port = uri.port.not_nil!  if uri.port
+        @db = uri.path.to_s.lchop if uri.path
+        @user = uri.user.to_s if uri.user
+        @password = uri.password.to_s if uri.password
+
+        if uri.query
+          params = HTTP::Params.parse(uri.query.to_s)
+          {% for field in CONNECTION_URI_PARAMS %}
+            {% if STRING_FIELDS.includes?(field) %}
+                @{{field.id}} = params["{{field.id}}"] if params["{{field.id}}"]?
+            {% end %}
+            {% if INT_FIELDS.includes?(field) %}
+                @{{field.id}} = params["{{field.id}}"].to_i if params["{{field.id}}"]?
+            {% end %}
+            {% if FLOAT_FIELDS.includes?(field) %}
+                @{{field.id}} = params["{{field.id}}"].to_f if params["{{field.id}}"]?
+            {% end %}
+          {% end %}
+        end
+        validate_config
+        self
+      end
 
-    def self.configure(&block)
-      yield self
-      self.validate_config
+      def read(path : String, env : String | Symbol = :development)
+        _env = env.to_s
+        source = YAML.parse(File.read(path))[_env]
+        {% for field in STRING_FIELDS %}
+          @{{field.id}} = source["{{field.id}}"].as_s if source["{{field.id}}"]?
+        {% end %}
+        {% for field in INT_FIELDS %}
+          @{{field.id}} = source["{{field.id}}"].as_s.to_i if source["{{field.id}}"]?
+        {% end %}
+        {% for field in FLOAT_FIELDS %}
+          @{{field.id}} = source["{{field.id}}"].as_s.to_f if source["{{field.id}}"]?
+        {% end %}
+        self.validate_config
+        self
+      end
     end
 
-    def self.validate_config
-      raise Jennifer::InvalidConfig.new("No adapter configured") if adapter.empty?
-      raise Jennifer::InvalidConfig.new("No database configured") if db.empty?
-    end
+    @@configs = {} of String => ConfigInstance
 
-    def self.configure
-      self
+    def self.get_instance(config_key : String | Symbol = :default)
```

Comment:
would be nice to just use symbols here but migrations will supply the key from the command line so the string is necessary.